modules/binary-tuple/src/main/java/org/apache/ignite/internal/binarytuple/inlineschema/TupleWithSchemaMarshalling.java

```diff
@@ -65,64 +71,105 @@ public final class TupleWithSchemaMarshalling {
             types[i] = inferType(value);
         }
 
-        BinaryTupleBuilder builder = builder(values, columns, types);
+        BinaryTupleBuilder schemaBuilder = schemaBuilder(columns, types);
+        BinaryTupleBuilder valueBuilder = valueBuilder(values, types);
+
+        byte[] schemaBt = schemaBuilder.build().array();
+        byte[] valueBt = valueBuilder.build().array();
+        // Size: int32 (tuple size), int32 (value offset), schema, value.
+        byte[] result = new byte[4 + 4 + schemaBt.length + valueBt.length];
 
-        byte[] arr = builder.build().array();
-        byte[] result = new byte[arr.length + 1];
-        // Put the size of the schema in the first byte.
-        result[0] = (byte) size; // todo should be int32
+        // Put the size of the schema in the first 4 bytes.
+        result[0] = (byte) (size >> 24);
```

Comment:
Alternatively, use `ByteBuffer`.

Reply:
Thanks, I've decided to copy methods from `HeapByteBufUtil`. The reason why I have not reused this class directly is that it has package-private methods and requires changes. But I do not want to modify this class since it is forked.